core/src/main/java/org/testcontainers/utility/DockerImageName.java

```diff
@@ -1,20 +1,173 @@
 package org.testcontainers.utility;
 
+
+import com.google.common.net.HostAndPort;
+import lombok.Data;
+import lombok.EqualsAndHashCode;
+
+import java.util.regex.Pattern;
+
+@EqualsAndHashCode
 public final class DockerImageName {
 
-    public static void validate(String dockerImageName) throws IllegalArgumentException {
-        int repoSeparatorIndex = dockerImageName.indexOf('/');
-        int tagSeparatorIndex;
-        if (repoSeparatorIndex == -1) {
-            tagSeparatorIndex = dockerImageName.indexOf(':');
+    /* Regex patterns used for validation */
+    private static final String ALPHA_NUMERIC = "[a-z0-9]+";
+    private static final String SEPARATOR = "([\\.]{1}|_{1,2}|-+)";
+    private static final String REPO_NAME_PART = ALPHA_NUMERIC + "(" + SEPARATOR + ALPHA_NUMERIC + ")*";
+    private static final Pattern REPO_NAME = Pattern.compile(REPO_NAME_PART + "(/" + REPO_NAME_PART + ")*");
+
+    private final String rawName;
+    private final String registry;
+    private final String repo;
+    private final Versioning versioning;
+
+    public DockerImageName(String name) {
+        this.rawName = name;
+        final int slashIndex = name.indexOf('/');
+
+        String remoteName;
+        if (slashIndex == -1 ||
```

Comment:
FWIW this logic, and the regex patterns, were copied/translated from an official Docker library - I didn't figure these out myself 😂